src/chainstate/stacks/index/storage.rs

```diff
@@ -965,6 +1646,69 @@ impl<'a, T: MarfTrieId> TrieStorageTransaction<'a, T> {
         Ok(ret)
     }
 
+    /// Run `cls` with a mutable reference to the inner trie blobs opt.
+    fn with_trie_blobs<F, R>(&mut self, cls: F) -> R
+    where
+        F: FnOnce(&Connection, &mut Option<&mut TrieFile>) -> R,
+    {
+        let mut blobs = self.blobs.take();
+        let res = cls(&self.db, &mut blobs);
+        self.blobs = blobs;
+        res
```

Comment:
Does this need to `take`?

```suggestion
        cls(&self.db, self.blobs.as_mut())
```

Reply:
It seems like any use of this method could just use `&self.db` and `self.blobs.as_mut()` directly, rather than needing a method with a function generic.

Reply:
Yes, it needs `.take()`.  You need to move `blobs` out of `self` so you can call a mutable method in `blobs` that takes an immutable reference to a member of `self`.  Otherwise you get borrow-checker errors:

```
   Compiling blockstack-core v0.0.1 (/home/jude/stacks/stacks-blockchain.marf-node-cache)
error[E0502]: cannot borrow `*self` as immutable because it is also borrowed as mutable
    --> src/chainstate/stacks/index/storage.rs:1759:63
     |
1758 |                     match self.blobs.as_mut() {
     |                           ---- mutable borrow occurs here
1759 |                         Some(blobs) => blobs.store_trie_blob(&self.db, &bhh, &buffer)?,
     |                                              ---------------  ^^^^ immutable borrow occurs here
     |                                              |
     |                                              mutable borrow later used by call

error[E0502]: cannot borrow `*self` as immutable because it is also borrowed as mutable
    --> src/chainstate/stacks/index/storage.rs:1784:63
     |
1783 |                     match self.blobs.as_mut() {
     |                           ---- mutable borrow occurs here
1784 |                         Some(blobs) => blobs.store_trie_blob(&self.db, real_bhh, &buffer)?,
     |                                              ---------------  ^^^^ immutable borrow occurs here
     |                                              |
     |                                              mutable borrow later used by call
```